File: backend/app/services/income.py
from decimal import Decimal

from sqlalchemy.orm import Session

from app.models.income import Income
from app.repositories.income import IncomeRepository
from app.schemas.income import IncomeCreate, IncomeList, IncomeOut, IncomeUpdate


class IncomeNotFoundError(Exception):
    pass


def list_income(
    db: Session,
    user_id: int,
    month: int | None = None,
    year: int | None = None,
) -> IncomeList:
    items = IncomeRepository.list_by_user(db, user_id, month=month, year=year)
    total = sum((item.amount for item in items), Decimal("0"))
    return IncomeList(
        items=[IncomeOut.model_validate(item) for item in items],
        total=total,
        count=len(items),
        month=month,
        year=year,
    )


def get_income(db: Session, income_id: int, user_id: int) -> Income:
    income = IncomeRepository.get_by_id_and_user(db, income_id, user_id)
    if income is None:
        raise IncomeNotFoundError("Income not found")
    return income


def create_income(db: Session, user_id: int, data: IncomeCreate) -> Income:
    return IncomeRepository.create(db, user_id, data)


def update_income(db: Session, income_id: int, user_id: int, data: IncomeUpdate) -> Income:
    income = get_income(db, income_id, user_id)
    return IncomeRepository.update(db, income, data)


def delete_income(db: Session, income_id: int, user_id: int) -> None:
    income = get_income(db, income_id, user_id)
    IncomeRepository.delete(db, income)

File: services/income.js
import { apiClient } from '@/services/api-client';

export async function getIncomes({ month, year }) {
  const { data } = await apiClient.get('/api/v1/income', { params: { month, year } });
  return data;
}

export async function getIncome(id) {
  const { data } = await apiClient.get(`/api/v1/income/${id}`);
  return data;
}

export async function createIncome(payload) {
  const { data } = await apiClient.post('/api/v1/income', payload);
  return data;
}

export async function updateIncome(id, payload) {
  const { data } = await apiClient.put(`/api/v1/income/${id}`, payload);
  return data;
}

export async function deleteIncome(id) {
  await apiClient.delete(`/api/v1/income/${id}`);
}
